### PR TITLE
Implements safe casting of to_<dtypes>() (#12235)

### PR DESCRIPTION
**Pull request by Edwin222**
* saturate_cast() supports SparseTensor

* to_float(), to_double(), to_int32(), to_int36(), to_bfloat16()
supports safe casting

### CODE REVIEW

**Comment by drpngx** [comment]
@ebrevdo is working on sparse support. Any opinion here?

**Comment by sb2nov** [comment]
@ebrevdo ping

**Comment by martinwicke** [comment]
@ebrevdo can you take a look?

File: tensorflow/python/ops/math_ops.py [hunk 1]
```diff
@@ -753,7 +752,7 @@ def saturate_cast(value, dtype, name=None):
   applies the appropriate clamping before the cast.
 
   Args:
-    value: A `Tensor`.
+    value: A `Tensor` or 'SparseTensor'.
```
Comment (ebrevdo) [inline tensorflow/python/ops/math_ops.py]: nit: `SparseTensor` (backticks)

File: tensorflow/python/ops/math_ops.py [hunk 2]
```diff
@@ -763,99 +762,126 @@ def saturate_cast(value, dtype, name=None):
...
 
 
-def to_double(x, name="ToDouble"):
+def to_double(x, name="ToDouble", mode="unsafe"):
   """Casts a tensor to type `float64`.
 
   Args:
     x: A `Tensor` or `SparseTensor`.
     name: A name for the operation (optional).
+    mode: "safe"/"unsafe" mode setting
```
Comment (ebrevdo) [inline tensorflow/python/ops/math_ops.py]: all these functions need more details about what safe and unsafe mean.

File: tensorflow/python/ops/math_ops.py [hunk 3]
```diff
@@ -763,99 +762,126 @@ def saturate_cast(value, dtype, name=None):
...
 
   Returns:
     A `Tensor` or `SparseTensor` with same shape as `x` with type `float64`.
 
   Raises:
     TypeError: If `x` cannot be cast to the `float64`.
   """
-  return cast(x, dtypes.float64, name=name)
+  if mode == "safe":
+    return saturate_cast(x,dtypes.float64, name=name)
```
Comment (ebrevdo) [inline tensorflow/python/ops/math_ops.py]: nit: space after x; more generally please review your python code throughout for spacing and other small nits.

**Review by ebrevdo — CHANGES_REQUESTED** [review]
all the changes need unit tests for safe mode.

also the new sparsetensor branch needs to be tested.

**Comment by martinwicke** [comment]
@Edwin222 any news? I'll mark this as stalled.

**Comment by drpngx** [comment]
Closing for now. Feel free to reopen when you're ready.

**Pull request closed without merging**